iceoryx2-services/discovery/src/service_discovery/service.rs

```diff
@@ -385,6 +453,43 @@ impl<S: ServiceType> Service<S> {
             }
         }
 
+        // Handle server requests
+        self.handle_discovery_requests()?;
+
+        Ok(())
+    }
+
+    /// Returns the service details of all the current services.
+    ///
+    /// This function is called within the spin function, so that
+    /// if any requests are recieved for the discovery states,
+    /// it can be provided via the RequestResponse method.
+    ///
+    /// # Returns
+    ///
+    /// A result containing `()` if successful.
+    ///
+    /// # Errors
+    ///
+    /// Returns a `ServerSpinError` if there was an error in Responsding,
+    /// Loaning or recieving Requests.
+    pub fn handle_discovery_requests(&mut self) -> Result<(), ServerSpinError> {
+        if let Some(server) = self.server.as_mut() {
+            while let Some(active_request) = server.receive()? {
+                let services = self.tracker.get_all();
+
+                for service in services.iter() {
+                    if !self.discovery_config.include_internal
+                        && ServiceName::has_iox2_prefix(service.name())
+                    {
+                        continue;
+                    }
+                    let response = active_request.loan_uninit()?;
+                    let response = response.write_payload(service.clone());
+                    response.send()?;
```

Comment:
When the response buffer is full this call would not be able to deliver it. When now two separate clients request a discovery result and on client does not receive it on purpose, it deadlocks the other client and this service.

Since iceoryx2 is intended for mission-critical system we must always ensure that a failure/misbehavior in one process does not affect any other process. In this case a rogue client could deadlock this service and prevent that any other client receives any more updates.

This must be handled differently.